server/service/src/item/bundled_item/test/mod.rs

```diff
@@ -0,0 +1,118 @@
+        service
+            .upsert_bundled_item(
+                &context,
+                UpsertBundledItem {
+                    id: test_bundled_item_record_id.to_string(),
+                    principal_item_variant_id: mock_item_a_variant_1().id,
+                    bundled_item_variant_id: mock_item_b_variant_1().id,
+                    ratio: 1.0,
+                },
+            )
+            .unwrap();
+
+        // Create a new bundled item for item_a_variant_2 to bundle with item_b_variant_2
+        // This is just to make sure that we can have multiple bundled items in the db
+        service
+            .upsert_bundled_item(
+                &context,
+                UpsertBundledItem {
+                    id: uuid(),
+                    principal_item_variant_id: mock_item_a_variant_1().id,
+                    bundled_item_variant_id: mock_item_b_variant_1().id,
+                    ratio: 1.0,
+                },
+            )
+            .unwrap();
+
+        // Update the ratio
+        let _bundled_item = service
+            .upsert_bundled_item(
+                &context,
+                UpsertBundledItem {
+                    id: test_bundled_item_record_id.to_string(),
+                    principal_item_variant_id: mock_item_a().id,
+                    bundled_item_variant_id: mock_item_a().id,
+                    ratio: 2.0,
+                },
+            )
+            .unwrap();
+
+        // Query the bundled item by id
+        let bundled_item = service
+            .get_bundled_items(
+                &context,
+                None,
+                Some(
+                    BundledItemFilter::new().id(EqualFilter::equal_to(test_bundled_item_record_id)),
+                ),
+            )
+            .unwrap();
+
+        assert_eq!(bundled_item.count, 1);
+        assert_eq!(
+            bundled_item.rows[0].id,
+            test_bundled_item_record_id.to_string(),
+        );
+        assert_eq!(bundled_item.rows[0].ratio, 2.0);
+
+        // Delete the bundled item
+        service
+            .delete_bundled_item(
+                &context,
+                DeleteBundledItem {
+                    id: test_bundled_item_record_id.to_string(),
+                },
+            )
+            .unwrap();
+
+        // Check that the delete worked
+        let bundled_item = service
+            .get_bundled_items(
+                &context,
+                None,
+                Some(
+                    BundledItemFilter::new().id(EqualFilter::equal_to(test_bundled_item_record_id)),
+                ),
+            )
+            .unwrap();
+
+        assert_eq!(bundled_item.count, 0);
+    }
+
+    #[actix_rt::test]
+    async fn validate_bundled_item() {
+        // TODO validation tests
+
+        // Check that the two item variants are not the same
+
+        // Check that the two item variants are not from the same item (that would be bad)
```

Comment:
Anything else to check here?

Reply:
I have two more:

- Can't bundle the same 2 variants multiple times (otherwise could configure same bundle with different ratios, which one should we pick?)
- Prevent nested bundling - check the `principal` variant isn't the `bundled` variant in another bundle